stdlib/universe/mode.go

```diff
@@ -0,0 +1,519 @@
+	s, ok := spec.(*ModeProcedureSpec)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	d := execute.NewDataset(id, mode, cache)
+	t := NewModeTransformation(d, cache, s)
+	return t, d, nil
+}
+
+type modeTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+
+	column string
+}
+
+func NewModeTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *ModeProcedureSpec) *modeTransformation {
+	return &modeTransformation{
+		d:      d,
+		cache:  cache,
+		column: spec.Column,
+	}
+}
+
+func (t *modeTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *modeTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("mode found duplicate table with key: %v", tbl.Key())
+	}
+
+	colIdx := execute.ColIdx(t.column, tbl.Cols())
+	if colIdx < 0 {
+		// doesn't exist in this table, so add an empty value
+		if err := execute.AddTableKeyCols(tbl.Key(), builder); err != nil {
+			return err
+		}
+		colIdx, err := builder.AddCol(flux.ColMeta{
+			Label: execute.DefaultValueColLabel,
+			Type:  flux.TString,
+		})
+		if err != nil {
+			return err
+		}
+
+		if err := builder.AppendString(colIdx, ""); err != nil {
+			return err
+		}
+		if err := execute.AppendKeyValues(tbl.Key(), builder); err != nil {
+			return err
+		}
+		// TODO: hack required to ensure data flows downstream
+		return tbl.Do(func(flux.ColReader) error {
+			return nil
+		})
+	}
+
+	col := tbl.Cols()[colIdx]
+
+	if err := execute.AddTableKeyCols(tbl.Key(), builder); err != nil {
+		return err
+	}
+	colIdx, err := builder.AddCol(flux.ColMeta{
+		Label: execute.DefaultValueColLabel,
+		Type:  col.Type,
+	})
+	if err != nil {
+		return err
+	}
+
+	if tbl.Key().HasCol(t.column) {
+		j := execute.ColIdx(t.column, tbl.Key().Cols())
+		switch col.Type {
+		case flux.TBool:
+			if err := builder.AppendBool(colIdx, tbl.Key().ValueBool(j)); err != nil {
+				return err
+			}
+		case flux.TInt:
+			if err := builder.AppendInt(colIdx, tbl.Key().ValueInt(j)); err != nil {
+				return err
+			}
+		case flux.TUInt:
+			if err := builder.AppendUInt(colIdx, tbl.Key().ValueUInt(j)); err != nil {
+				return err
+			}
+		case flux.TFloat:
+			if err := builder.AppendFloat(colIdx, tbl.Key().ValueFloat(j)); err != nil {
+				return err
+			}
+		case flux.TString:
+			if err := builder.AppendString(colIdx, tbl.Key().ValueString(j)); err != nil {
+				return err
+			}
+		case flux.TTime:
+			if err := builder.AppendTime(colIdx, tbl.Key().ValueTime(j)); err != nil {
+				return err
+			}
+		}
+
+		if err := execute.AppendKeyValues(tbl.Key(), builder); err != nil {
+			return err
+		}
+		// TODO: hack required to ensure data flows downstream
+		return tbl.Do(func(flux.ColReader) error {
+			return nil
+		})
+	}
+
+	var (
```

Comment:
Instead of defining all of the various types in the outermost portion, I think it would be good if we just determined the column type and then went to a function to deal with that column type instead of having to type switch on the same type repeatedly. I could go into more detail, but the aggregate interface spells it out better than I think I would be able to in a short paragraph. Could you take a look at that?

Reply:
followed distinct.go here because we ended up deciding on a custom OpSpec vs. aggregate

Reply:
I think regardless of whether or not this uses the aggregate functions, it should still be using a similar pattern. The method with distinct is very difficult to read so I would like to have a struct definition for each of the types that takes the same signature as aggregate and uses the same method, regardless of whether it uses the exact code because of returning multiple values.